pkg/storage/unified/resource/server.go

```diff
@@ -581,21 +622,37 @@ func (s *server) Watch(req *WatchRequest, srv ResourceStore_WatchServer) error {
 			}
 
 			if event.ResourceVersion > since && matchesQueryKey(req.Options.Key, event.Key) {
-				// Currently sending *every* event
-				// if req.Options.Labels != nil {
-				// 	// match *either* the old or new object
-				// }
-				// TODO: return values that match either the old or the new
-
-				if err := srv.Send(&WatchEvent{
+				value := event.Value
+				// remove the delete marker stored in the value for deleted objects
+				if event.Type == WatchEvent_DELETED {
+					value = []byte{}
```

Comment:
Yes. This part happens on the client side [here](https://github.com/grafana/grafana/blob/d1a299016628dd56a32b96d6c50d23702d4837e6/pkg/storage/unified/apistore/stream.go#L163-L179)

Why does it happen on the client side ? because the actual event type sent to the client depends on the list options. For example, If the user has a label (or field) selector, then and update event that was matching on the old object might not match on the new object resulting in a Delete event.